hw/dv/sv/dv_lib/dv_base_env.sv

```
@@ -32,11 +32,11 @@ class dv_base_env #(type CFG_T               = dv_base_env_cfg,
 
         if (ral_name == default_ral_name) if_name = "clk_rst_vif";
         else                              if_name = {"clk_rst_vif_", ral_name};
-        if (!uvm_config_db#(virtual clk_rst_if)::get(this, "", if_name,
-            cfg.clk_rst_vifs[ral_name])) begin
-          `uvm_fatal(`gfn, $sformatf("failed to get clk_rst_if for %0s from uvm_config_db", ral_name))
+        if (uvm_config_db#(virtual clk_rst_if)::exists(this, "", if_name)) begin
+          uvm_config_db#(virtual clk_rst_if)::get(this, "", if_name,
+              cfg.clk_rst_vifs[ral_name]);
+          cfg.clk_rst_vifs[ral_name].set_freq_mhz(cfg.clk_freqs_mhz[ral_name]);
```

we need to obtain a vif for each element of the cfg.clk_rst_vifs[], right? if it doesn't exist, should report a fatal error.

---

I wasn't sure what to do where there is only one clock and reset, but two RALs. Hence making it optional to have a clk_rst_vif for every RAL (there is still a check below that the default clk_rst_vif gets assigned).

I'll have another go at getting it to work with a second dummy clk_rst_if in the tb to see if that's cleaner.

---

sounds good, thanks

---

@weicaiyang I just force-pushed an alternative version of the `rom_ctrl` TB which creates the second `clk_rst_vif`.

The only issue with this approach is that it leaves a bit of a delay between the dut reset releasing and the `apply_reset` task completing (since there is a second reset being applied to the dummy `clk_rst_vif` which can have a longer delay than the main `clk_rst_vif`) but I have worked around that in the TB.

---

rethinking this. the dummy clock is also a bit strange. I didn't thinking of same clock for multiple RALs at the beginning. 
We actually don't use clk_rst_vifs in CSR automatic tests. It's not a bad idea to only keep the actual interfaces in clk_rst_vifs, user just need  to keep in mind we don't have one-to-one mapping for RAL and vif, and use the correct vif manually.

the dummy clock solution also works too. In some seeds, it has some extra delay after reset, which should be fine. So, I think either way is fine to me.